apps/local_laplacian/local_laplacian_generator.cpp

```diff
@@ -124,17 +117,16 @@ class LocalLaplacian : public Halide::Generator<LocalLaplacian> {
             }
             outGPyramid[0]
                 .compute_at(output, y).vectorize(x, 8);
-            for (int j = 5; j < J; j++) {
+            for (int j = 5; j < pyramid_levels; j++) {
                 inGPyramid[j].compute_root();
                 gPyramid[j].compute_root().parallel(k);
                 outGPyramid[j].compute_root();
             }
         }
-
-        return output;
     }
 private:
     Var x, y, c, k;
+    Func remap, gray, inGPyramid[maxJ], gPyramid[maxJ], outGPyramid[maxJ];
```

Comment:
I have an idea:

The reason we separate schedule() from generate() is so that scheduling can be deferred, in case it depends on symbols that come from later in the algorithm definition (e.g. a compute_at LoopLevel). 

But the whole essence of the algorithm language is deferred execution -- you can use a Param<int> in your algorithm without knowing its value yet.

So what if we allowed scheduling to use unbound ScheduleParams? What if ScheduleParams were a little more like Params and less like GeneratorParams? They all need to be bound before you actually compile, but you can do that at any point up until lowering. There would be a phase at the top of lowering that resolved the necessary level of indirection. (e.g. a compute_at(schedule_param_5) would become a compute_at(schedule_param_5's bound value)).

ScheduleParams in scalar types would act as Exprs that get replaced by constant values at the top of lowering, so you can tile using them, specialize on them to turn on/off vectorization, etc.

Usage would be something like:

    auto demosaic = MyDemosaicGeneratorOrStub::do_something( ... inputs ...);

    // define rest of pipeline ...
     ... demosaic.output(x, y) ...

    // schedule for processed
    ....

    // schedule for demosaic
    demosaic.compute_at_level = {processed, y}
    demosaic.store_at_level = {processed, yo}

    // Always vectorize demosaic in this context. Other uses of demosaic might not.
    demosaic.vector_lanes = 8;
    // I've bound the scheduleparam to a constant.
    // Demosaic already ran code like foo.specialize(vector_lanes > 1).vectorize(x, vector_lanes);

    // parallelize the demosaic only when I'm not being parallelized
    demosaic.parallelize = !(this->parallelize);  
    // I've bound the scheduleparam to an Expr that depends on one of my scheduleparams

Note that the last bit assumes you can bind scheduleparams using any Expr, including other unresolved scheduleparams, which allows transitively communicating information upwards.